### ISSUE TITLE
connect() overwrites local data and disconnect() clears local data, why?

### ISSUE THREAD

**Issue by dienluong** [remotestorage/remotestorage.js#1170]
Hello,

First, thank you for the project!

rs.js is advertised as "offline-first by design".  But how can an app be offline-first when local data is not kept?

To give a simple example: let's say we have a notes taking app.  When not connected to a remote storage, the user can still create, delete and update notes.  Once connected to online storage, the new or modified notes get sync'd with the data in online storage.  That's not possible with rs.js because, based on my testing, when you connect, the online version of your data overwrites the local copy (i.e. your local modifications are lost).  Furthermore, when the user disconnects from the online storage, the local copy is destroyed.  So the user can no longer work on the existing notes, when disconnected.

So it seems to me that rs.js, despite the claimed "offline-first design", does not enable an app to be offline-frst.

Update: I think I realize now what "offline-first design" means in the context of rs.js.  It means no Internet/remote connection but still "**logged in**" to a remote storage account (as opposed being in "anonymous mode").

**Comment by skddc** [comment]
Hi Dien,

Thanks for your question. The answer to this is two-fold, and after testing this for a bit in one of my apps, it seems you may have found a bug in the library. Let me explain:

### Connecting a remote storage with existing local data

Data created or edited while being offline/disconnected should *not* overwrite remote data by default. When connecting a remote storage, the data should be synced exactly the same way as if your app had been connected, but temporarily lost Internet connectivity.

The way the sync works is generally "last write wins", but the client will know if the same object/file has been modified on the server since last synchronizing it (by way of storing the ETAG), and in case the file has been modified both in the local and remote storage, a `conflict` event should be emitted, which apps can then use to resolve such conflicts.

When I tested this in [Webmarks](https://webmarks.5apps.com/) just now, it worked fine with new data/bookmarks, but there seems to be a bug when adding the same bookmark before connecting. After the sync, I can see the changes in the UI, but when reloading the page, or re-connecting the storage, it turns out that the changes haven't actually been synced. This can either be a bug in Webmarks or remoteStorage.js. I will investigate further to find out what is happening exactly.

### Disconnecting a remote storage

When disconnecting a remote storage from an app, the local cache is purged intentionally. It has been designed this way, because the most common use case for disconnecting a storage for us was usually wanting to connect a different account and/or to clear the local cache for other reasons.

However, I can see why you may want to keep the local data around, even while it's not synchronizing a remote storage. In fact, I think this could be used for easy data migration between accounts, provided the app developer has implemented full caching, as well as designed the storage usage in a way that doesn't break app data in case the data from two different accounts is mixed randomly upon such a sync procedure.

Perhaps it makes sense to explore this a bit further, and potentially provide a library and widget option for app developers and users to be able to choose not to purge the cache on disconnect.

**Comment by skddc** [comment]
Just FYI: added the docs label, because it seems like we should document the current and intended behavior better.

**Comment by dienluong** [comment]
Hi Sebastian,

Thank you for the response. Very appreciated.

> When disconnecting a remote storage from an app, the local cache is purged intentionally. It has been designed this way, because the most common use case for disconnecting a storage for us was usually wanting to connect a different account and/or to clear the local cache for other reasons.

Perhaps purging the local cache only upon connecting to a different account, rather than purging right after disconnecting from the current account?

Regards!

**Comment by galfert** [comment]
I don't think that's a good idea, at least not as a default behavior. When using an app on any other than your own computer or device, disconnecting your account would still keep your data on that machine.

**Comment by dienluong** [comment]
You are right. (I tend to forget about this use case, which rarely applies
to me, i.e. using shared desktop account or devices.)

**Comment by raucao** [comment]
FWIW, the bug mentioned above was just Webmarks not handling conflicts properly. All remaining sync issues that were still open have been fixed in #1331.

I guess for the docs, it would be best to add a "Disconnecting a storage" section to https://remotestorage.io/rs.js/docs/getting-started/ (analog to the "Connecting a storage" one). The [`disconnect()` doc](https://remotestorage.io/rs.js/docs/api/remotestorage/classes/RemoteStorage.html#disconnect) is already explicit and unambiguous about clearing the local cache.

PRs welcome! The Getting Started docs are written in Markdown ([sources here](https://github.com/remotestorage/remotestorage.js/tree/master/docs/getting-started)), and the `Caching` class docs can be updated directly in [src/caching.ts](https://github.com/remotestorage/remotestorage.js/blob/master/src/caching.ts#L6-L44).

**Comment by jonocodes** [comment]
> FWIW, the bug mentioned above was just Webmarks not handling conflicts properly. All remaining sync issues that were still open have been fixed in [#1331](https://github.com/remotestorage/remotestorage.js/pull/1331).


Ok. But webmarks is only one app experiencing this.
And the PR mentioned here could be a good direction:
https://github.com/remotestorage/remotestorage.js/pull/1237

I am musing 2.0.0-beta8 so I believe I have the latest bug fixes which do not include a change in this disconnect behavior.


> I guess for the docs, it would be best to add a "Disconnecting a storage" section to https://remotestorage.io/rs.js/docs/getting-started/ (analog to the "Connecting a storage" one). The [`disconnect()` doc](https://remotestorage.io/rs.js/docs/api/remotestorage/classes/RemoteStorage.html#disconnect) is already explicit and unambiguous about clearing the local cache.


To clarify the current behavior, disconnecting causes the local data to be removed? And this is by design. Can you clarify what the 'cache' is for if its not to support disconnected remote storage? Is it mostly for loss of network connectivity?


> PRs welcome! The Getting Started docs are written in Markdown ([sources here](https://github.com/remotestorage/remotestorage.js/tree/master/docs/getting-started)), and the `Caching` class docs can be updated directly in [src/caching.ts](https://github.com/remotestorage/remotestorage.js/blob/master/src/caching.ts#L6-L44).

Sounds good about contributing. Once I wrap my head around some of this.

**Comment by jonocodes** [comment]
I dug into #1237 and the cleanup code. The reason that PR got rejected is real: skipping all _cleanups would leave the bearer token in localStorage and leave sync running with its listeners. The cache-wiping is only two cleanups out of the seven — indexeddb and localstorage. Everything else (sync teardown, token removal, backend unhook) has to keep running on disconnect regardless.

So a flag that gets read by just those two cleanups, not a global bypass, is the shape that works.

The harder question is multi-user. If we preserve the IDB across disconnect, reconnecting under a different userAddress reads the previous user's cached blobs until sync overwrites them path-by-path — and orphan paths stick around forever. I looked at namespacing the DB by userAddress to avoid this, but the DB name is enumerable via indexedDB.databases() and visible in DevTools, so putting an email there leaks PII to any same-origin script. Hashing also isnt great, but it could work.

The version I think works: keep the single DB, but store a marker of which userAddress the cache belongs to. On connect, if it matches, reuse the cache; if it doesn't, wipe before proceeding. You lose the warm cache when switching between two users on one browser, but that's rare and the dominant case (one user, personal device, occasional disconnect) gets what it wants. No PII in any metadata.

So something like:

```
new RemoteStorage({ preserveLocalDataOnDisconnect: true });
remoteStorage.disconnect({ preserve: true });   // per-call override
remoteStorage.deleteLocalData();                // explicit wipe later
Default stays false, so galfert's shared-device case is unchanged.
```

For anyone hitting this today: monkey-patching featureModules.IndexedDB._rs_cleanup to call closeDB() without IndexedDB.clean() works as a workaround if you're single-user. This is the solution I patched in for my app to preserve the data: https://github.com/jonocodes/savr/commit/f501103f846eb8e8729343c968833f44e73e55ce#diff-9c4b295fe8cbfba60323f5e515db6befde076ef7a041d368ff96201518339afd

@raucao does this match what you were picturing?

**Comment by raucao** [comment]
Yes, this is more workable. However, @galfert had started working on true multi-account support (have multiple accounts connected at the same time, e.g. for a data migration app), and only adding metadata to the current cache would still keep it a single-account solution.

So I think it would make more sense to get most of the required functionality for this feature as a side effect of implementing multi-account connections/sync.

**Comment by jonocodes** [comment]
Ok, so it sounds like this is sort of blocked by that multi-account work. @galfert could you like the progress or ticket so we can track that work here. Thanks.

**Comment by raucao** [comment]
Not blocked per se. Just saying that, in my opinion, we should probably go in that direction instead, because it would solve most of what's required for disconnected caches as well, in addition to enabling multi-account connections (think e.g. personal and business account in the same to-do app, etc.).
